### PR TITLE
[Core] [runtime env] Avoid spurious worker startup

### LINKED ISSUE #16591: [runtime_env] Avoid spurious worker startup while envs are installing

### PR DESCRIPTION
**Pull request by architkulkarni**
## Why are these changes needed?
While a runtime env is installing (which may take a minute or so), `PopWorker()` is continually called every few seconds for the same task.  This leads to unnecessary extra workers being started.  This PR adds the following quick fix: if a runtime env is currently being installed, then PopWorker returns `nullptr` as before, but does *not* spawn a new worker.

## Related issue number

Closes #16591 
## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: src/ray/raylet/worker_pool.cc
Comment (architkulkarni) [inline src/ray/raylet/worker_pool.cc:1003]: @edoakes this approach with `pending_runtime_envs_` looks different from the approach we discussed with the map to states (`PENDING`, `DONE`) but I think the outcome is the same--just wanted to double check that I wasn't overlooking something.

File: src/ray/raylet/worker_pool.cc
Comment (edoakes) [inline src/ray/raylet/worker_pool.cc:995]: This comment could use some more context

File: src/ray/raylet/worker_pool.cc
Comment (edoakes) [reply src/ray/raylet/worker_pool.cc:1003]: I think the behavior here is slightly different than what we discussed previously. With the current implementation, we're effectively throttling it so that we can only ever have a single outstanding RPC to the agent. In the case where we've already created the env, we might want to start up a number of workers quickly and with this implementation we'll be doing it serially.

This behavior might be OK, but it would be a different from what we have with the non-runtime env case and I'd prefer to keep it the same if possible.

File: src/ray/raylet/worker_pool_test.cc
Comment (edoakes) [inline src/ray/raylet/worker_pool_test.cc:1296]: Check that the worker_pool size is still zero here?

File: src/ray/raylet/worker_pool_test.cc
Comment (edoakes) [inline src/ray/raylet/worker_pool_test.cc:1305]: We should also check that after this we can create a second worker with the same env.

You might consider adding a test that multiple RPCs will or will not be sent in parallel depending on the resolution from the comment above about this approach.

File: src/ray/raylet/worker_pool.cc
Comment (architkulkarni) [reply src/ray/raylet/worker_pool.cc:1003]: Ah good point, let me go back to the other approach.

**Review by rkooo567 — COMMENTED** [review]
Hey @architkulkarni can you also add a python test as well?

**Review by edoakes — COMMENTED** [review]
LGTM pending python test

File: src/ray/raylet/worker_pool.cc
Comment (simon-mo) [inline src/ray/raylet/worker_pool.cc:1019]: I think info still go to the driver? can you double check to make sure this doesn't propagate to the driver?

**Review by simon-mo — APPROVED** [review]
LGTM

File: src/ray/raylet/worker_pool.cc
Comment (architkulkarni) [reply src/ray/raylet/worker_pool.cc:1019]: Good point, it turns out it doesn't go to the driver.  I checked the logs and saw INFO logs from worker_pool.cc in raylet.out, but they never showed up in the driver.

**Comment by rkooo567** [comment]
@architkulkarni sorry, but can we actually go through @wuisawesome's review one last time before merging this?

File: src/ray/raylet/worker_pool.cc
Comment (wuisawesome) [reply src/ray/raylet/worker_pool.cc:1019]: Yup, INFO from raylet shouldn't go to driver. That being said, it might be better to make the DEBUG so it doesn't eat up all the disk space on a long running cluster.

File: src/ray/raylet/worker_pool.cc
Comment (wuisawesome) [inline src/ray/raylet/worker_pool.cc:1013]: I realize this was just copied/pasted from before, but can we include some more info in the error message (and also make it a warning since tasks can be retried by the end user)?

```
RAY_LOG(WARNING) << "Couldn't create a runtime environment for task " << task_spec.TaskId() << ". The runtime environment was " << task_spec.SerializedRuntimeEnv() << ".";
```

**Review by wuisawesome — APPROVED** [review]
Thanks! Btw have we confirmed this works with the many actors release test?

File: src/ray/raylet/worker_pool.cc
Comment (architkulkarni) [reply src/ray/raylet/worker_pool.cc:1019]: Good point, I didn't think of that

**Comment by architkulkarni** [comment]
> Thanks! Btw have we confirmed this works with the many actors release test?
We haven't, should I run it before merging this?  Also, I still need to add a Python test for this before merging.

**Comment by wuisawesome** [comment]
Not required, but it would be nice since this is obviously a section of code that has caused scalability issues in the past

**Comment by edoakes** [comment]
@wuisawesome how about we lean on the periodic testing here and monitor it after the change lands in master? I think this is pretty unlikely to cause a regression.

**Comment by wuisawesome** [comment]
sgtm

**Review by rkooo567 — CHANGES_REQUESTED** [review]
Sorry for the last blocker, but can you address this comment? https://github.com/ray-project/ray/pull/17422#pullrequestreview-718754291 (or justify why python test is not possible)?

**Comment by architkulkarni** [comment]
Oh no worries, I've been working on adding the Python test but just haven't finished yet.  Ed and I discussed an approach which doesn't seem too complicated.

**Comment by edoakes** [comment]
@rkooo567 - @architkulkarni and I looked into writing a python integration test here but it looks like it would be very tricky. We basically need some way to keep a log of python workers starting up without them actually executing anything. We could probably accomplish this by adding some sort of hook. The other option would be to record that the worker was started in either the env agent or the raylet and be able to query that info, but that's a roundabout way of doing essentially the same as what we're already doing in the unit tests.

Given that this wouldn't really test any additional behavior beyond what we're already covering in the unit tests, I think we should merge this without python tests.

**Comment by wuisawesome** [comment]
I think @stephanie-wang had some ideas about how to assert the number of workers that started up. There's a somewhat relevant discussion here: https://github.com/ray-project/ray/pull/17202#issuecomment-889373980

**Comment by rkooo567** [comment]
Maybe I am missing something, but isn’t this possible in this way? 

- start a cluster
- Schedule 2 tasks that has long artificial delay to process startup (we can mock runtime env?)
- and check the debug_state.txt to make sure only 2 worker process has started

**Comment by architkulkarni** [comment]
Ah thanks, I didn't know about `debug_state.txt`.  That sounds like exactly what we need, let me take a look.

**Review by rkooo567 — APPROVED** [review]
This is nice!

**Comment by rkooo567** [comment]
All the mac build didn’t seem to run. Can you retry with the latest master?

**Comment by architkulkarni** [comment]
Only failing test is //python/ray/tests:test_basic_3_client_mode which is known flaky on master.

**Pull request merged**
